api/turing/api/ensemblers_api.go

```diff
@@ -109,18 +111,162 @@ func (c EnsemblersController) UpdateEnsembler(
 			"Ensembler type cannot be changed after creation")
 	}
 
+	pyFuncEnsembler, isPyfunc := ensembler.(*models.PyFuncEnsembler)
+
+	var oldPyFuncEnsembler models.PyFuncEnsembler
+	if isPyfunc {
+		oldPyFuncEnsembler = models.PyFuncEnsembler{
+			RunID:       pyFuncEnsembler.RunID,
+			ArtifactURI: pyFuncEnsembler.ArtifactURI,
+		}
+	}
+
 	if err = ensembler.Patch(request.EnsemblerLike); err != nil {
 		return BadRequest("invalid ensembler configuration", err.Error())
 	}
-
 	ensembler, err = c.EnsemblersService.Save(ensembler)
 	if err != nil {
+		if isPyfunc {
+			_ = c.MlflowService.DeleteRun(context.Background(), pyFuncEnsembler.RunID, pyFuncEnsembler.ArtifactURI, true)
+		}
 		return InternalServerError("failed to update an ensembler", err.Error())
 	}
 
+	// Delete If Only RunID is changed, of only name is changed, doesn't need to delete the ensembler run
+	if isPyfunc {
+		_ = c.MlflowService.DeleteRun(context.Background(), oldPyFuncEnsembler.RunID, oldPyFuncEnsembler.ArtifactURI, true)
```

Comment:
This were the deletion of the old mlflow run, if the update already succeed, we need to delete the old mlflow run, and if the deletion were failed, to avoid confusion we just let it be, since if we show an error message, but the update were succeed, it will confuse the user

Reply:
Hmm.. We discussed something similar with the mlflow run deletion earlier, around the deletion of the GCS artifact. I'm of the opinion that the nested artifact should be deleted first and errors should block further processing. So in this case, can we delete the old run first before saving the new one in the Turing DB? If deletion of the old run failed, we can just show error to the user. Wdyt?

Reply:
if we delete the old mlflow run before updating the db, and when the db update failed, then the ensembler can't be used again since the mlflow run + gcs artifact is already deleted, that's the main reason why i put the deletion of the old ensembler last